src/main/java/com/google/devtools/build/lib/rules/proto/BazelProtoLibrary.java

```diff
@@ -33,15 +37,53 @@
   public ConfiguredTarget create(RuleContext ruleContext)
       throws ActionConflictException, RuleErrorException, InterruptedException {
     ProtoCommon.checkRuleHasValidMigrationTag(ruleContext);
-    ProtoInfo protoInfo =
-        ProtoCommon.createProtoInfo(
-            ruleContext,
-            ruleContext.getFragment(ProtoConfiguration.class).generatedProtosInVirtualImports());
+    ProtoCommon.checkSourceFilesAreInSamePackage(ruleContext);
+
+    ImmutableList<Artifact> originalDirectProtoSources =
+        ruleContext.getPrerequisiteArtifacts("srcs", Mode.TARGET).list();
+    ImmutableList<ProtoInfo> deps = ImmutableList.copyOf(
+            ruleContext.getPrerequisites("deps", TARGET, ProtoInfo.PROVIDER));
+    ImmutableList<ProtoInfo> exports = ImmutableList.copyOf(
+            ruleContext.getPrerequisites("exports", TARGET, ProtoInfo.PROVIDER));
+    String protoSourceRoot =
```

Comment:
What are you planning to add to `ProtoCommon`? Depending on the magnitude of the feature, we may have to get bazel-discuss / bazel-dev involved.

This will not be very satisfying, but the reason why the code is structured this way at HEAD is that we have a Google-internal alternative implementation of `BazelProtoLibrary` (which we want to go away, but it's still there) and we found it useful to have all the duplicate code in `ProtoCommon`. That's why `BazelProtoLibrary` is so short. 

Mind moving all the code you put here into a method in `ProtoCommon` that's separate from `createProtoInfo()` instead? (preferably after discussing your future plans)

Reply:
I want to add a `create_proto_info` interface to `ProtoCommon` so Starlark rules can create ProtoInfo providers.  This would let us build custom Proto rules that work with the existing proto rules.  It would also let use pull the rules out to https://github.com/bazelbuild/rules_proto

For the Google-internal alternative would it work to leave the current createProtoInfo(ruleContext) method in place and add a second createProtoInfo method without the ruleContext from this PR and have the createProtoInfo(ruleContext) call the new createProtoInfo method?

Reply:
Having two `createProtoInfo()` versions is reasonable, as long as one calls the other.

Reply:
I like the idea of providing a way to create `ProtoInfo` from Starlark. We'll definitely need it sooner-or-later to rewrite the proto rules in Starlark and move them to the `rules_proto` repo.

@cheister It seems like you already have something in mind. Would you mind creating a "design-doc issue" (either here or `rules_proto`) to see what the API will look like (i.e. is it going to be a simple constructor-like function, will it be more of a Sandwich API, ...). We can always involve the mailing list(s) later (if necessary).

Reply:
Sure.  Do you mind if I do that with the follow-up PR that includes the `create_proto_info` method?   What I had in mind was 
```
  ProtoInfoT createProtoInfo(
      String protoSourceRoot,
      FileApi descriptorSet,
      Sequence<?> sources,
      Sequence<?> deps,
      Sequence<?> exports,
      StarlarkThread thread)
      throws EvalException;
}
```

Reply:
TBH, I prefer an issue.
I'd also like to learn a little about use-cases for this API and maybe have an example of how usage will look like, which, IMO, is harder if we have a PR and look at the code. I absolutely don't expect more than a short write-up.

Reply:
+1. Given our previous thread about the inconsistencies and consistency verification / creation in such a way that inconsistency is not possible, I think it makes sense to break the thread out into an issue. No need for anything formal, just a place where we can decide on the API.